packages/driver/src/cy/commands/request.js

```diff
@@ -277,7 +277,23 @@ module.exports = (Commands, Cypress, cy, state, config) => {
       // because we're handling timeouts ourselves
       cy.clearTimeout('http:request')
 
-      return Cypress.backend('http:request', requestOpts)
+      return Promise.try(() => {
+        // https://github.com/cypress-io/cypress/issues/6178
+        // Check if body is Blob.
+        // construct.name is added because the parent of the Blob is not the same Blob
+        // if it's generated from the test spec code.
+        if (requestOpts.body instanceof Blob || requestOpts?.body?.constructor.name === 'Blob') {
```

Comment:
What about `File` or `ArrayBuffer`? Do we only seek to support `Blob`?

The websocket between the driver and server uses socket.io, and if you check out socket.io's source code, it supports transmitting `ArrayBuffer`, views, `File`, and `Blob`: https://github.com/socketio/socket.io-parser/blob/af1b23ca85fb64a0d7a050abd0362c5f632ce429/lib/is-binary.ts#L25-L31

Since `socket.io` already supports the serialization, maybe we should not `JSON.stringify` it at all before passing it to the websocket, and instead we should fix `Cypress.backend` to trust `socket.io`'s serialization?

What do you think about all this?

Reply:
Oh, look what I found @sainthkh: #1647 Request to support uploading formdata/File here

Reply:
I read the code of `socket.io` and learned that the `stringify` process is done by `socket.io`, not Cypress. 

The user-given data is sent to server with `runner` below:

https://github.com/cypress-io/cypress/blob/1e7a3faa4c26bea5aea12be4f5eb023235c3b1a4/packages/runner/src/lib/event-manager.js#L367-L371

When you check the `args` above, `body` is still `Blob` object.

And the data passed to `socket.io` is passed to [`socket.io`'s `emit()`](https://github.com/socketio/socket.io/blob/master/lib/socket.ts#L191-L217) to [`packet()`](https://github.com/socketio/socket.io/blob/master/lib/socket.ts#L281-L288) to [`socket.io-client`'s `_packet()`](https://github.com/socketio/socket.io-client/blob/2edf0f308c8b969928124d02e81e21c456f736a7/lib/manager.ts#L672-L679), which calls the `encode()` of [`socket.io-parser`'s `encode()`](https://github.com/socketio/socket.io-parser/blob/af1b23ca85fb64a0d7a050abd0362c5f632ce429/lib/index.ts#L44-L57) below:

```js
  public encode(obj: Packet) {
    debug("encoding packet %j", obj);


    if (obj.type === PacketType.EVENT || obj.type === PacketType.ACK) {
      if (hasBinary(obj)) { // <= this line
        obj.type =
          obj.type === PacketType.EVENT
            ? PacketType.BINARY_EVENT
            : PacketType.BINARY_ACK;
        return this.encodeAsBinary(obj);
      }
    }
    return [this.encodeAsString(obj)];
  }
```

It calls `hasBinary()` and it calls the `isBinary` function you referenced above. And I checked that [`hasBinary`](https://github.com/socketio/socket.io-parser/blob/af1b23ca85fb64a0d7a050abd0362c5f632ce429/lib/is-binary.ts#L33-L66) ignores JavaScript objects that contain binary data as in our case. It only checks the entire data is binary or not. I'm not sure why they ignored this case, but I suspect it's because of the performance. As we all know, traversing tree is really expensive. 

As for our case, I guess we need to mimic the behavior of `encodeAsBinary()`. It might not be hard because we can reuse the functions in `socket.io-parser`.

Phew, this issue seems much much bigger than I first thought.

Reply:
After server gets the data, I didn't do anything and it encoded it to string. To conserve it as binary, we need to set `encoding` to `null`. [Check this.](https://stackoverflow.com/questions/37703518/how-to-post-binary-data-in-request-using-request-library)

# TODO

- [ ] Fix test for `Blob`.
- [ ] Create tests for `ArrayBuffer`, `File`
- [ ] Send binary data to Cypress server.
- [ ] Send binary data to the user server.

Reply:
Wow, this is a lot more complicated than I thought. Great deep-dive.

If it makes more sense to split this into multiple PRs for Blob/File/formdata, just let me know, we can do that too.